### ISSUE TITLE
How to implement hierarchical DQN?

### ISSUE THREAD

**Issue by Trinkle23897** [thu-ml/tianshou#403]
@Trinkle23897 Hi there. I've got a similar problem with sparse reward and I'd like to try hierarchical reinforcement learning. But I have no idea how to implement a hierarchical RL algorithm like hierarchical DQN with tianshou. Could you please provide me with any instructions or example?

_Originally posted by @pengpeng1997 in https://github.com/thu-ml/tianshou/issues/237#issuecomment-886417062_

**Comment by Trinkle23897** [comment]
Here's one possible way (but a little bit tricky):

1. we have 2 policies, one for the low-level controller and one for meta-controller; thus there are two buffers, one store normal timestep (but with intrinsic reward), and the other one store goal timestep (with extrinsic reward);
2. add a gym.Wrapper: change state to `np.concatenate([state, goal])`, and the goal generates at every `env.reset()` or reach goal in `env.step()`. Modify the reward to two terms: intrinsic reward and extrinsic reward. Add the goal transition when one episode finishes or reach goal. (see the code below)
3. modify (or create a new class that inherit existing Policy class if you don't want to change the internal code) `policy.update` function to 
```python
def update(self, sample_size, buffer):
  self.meta_policy.update(sample_size, self.meta_buffer)  # for meta controller
  super().update(self, sample_size, buffer)  # for low-level controller
```

Env wrapper:
```python
class HRLWrapper(gym.Wrapper):
  def __init__(self, env, env_id, meta_policy, meta_buffer, intrinsic_reward_fn):
    super().__init__(env)
    self.meta_policy = meta_policy
    self.meta_buffer = meta_buffer
    self.env_id = env_id
    self.intrinsic_reward_fn = intrinsic_reward_fn
    self.goal = None
    self.observation_space = ...

  def new_goal(self, state):
    if self.goal is not None:
      self.meta_buffer.add(Batch(obs=[self.meta_state], act=[self.goal], rew=[self.extrinsic_reward],
                                 done=[self.done], obs_next=[self.state], info=[self.info]),
                           buffer_ids=[self.env_id])  # if you use multiple env
    self.extrinsic_reward = 0
    self.state = state
    self.goal = self.meta_policy(Batch(obs=self.state)).act[0]
    self.onehot_goal = F.one_hot(torch.tensor(self.goal), num_classes=self.action_space.n)

  def reset(self):
    obs = super().reset()
    self.new_goal(obs)
    self.obs = np.concatenate([self.state, self.onehot_goal])
    return self.obs

  def step(self, action):
    obs_next, rew, self.done, self.info = super().step(action)
    self.obs = np.concatenate([obs_next, self.onehot_goal])
    self.extrinsic_reward += rew
    intrinsic_reward = self.intrinsic_reward_fn(obs_next, self.goal)
    if obs_next == self.goal:
      self.new_goal(obs_next)
    return self.obs, intrinsic_reward, self.done, self.info
```
